rbreslow/zs
# Store gamemode content in S3

## Overview

- Move Garry's Mod sources under `src/garrysmod` to accommodate future `src/` modules.
- Remove `GM:AddResources()` and include autogenerated `sv_resources.lua`.
- Add `.dockerignore file` to avoid shipping large files (BSPs, textures, etc) to Docker build context.
- Add `awscli` Docker Compose service for interacting with S3.
- Add `scripts/sync-fastdl` for syncing FastDL artifacts to an S3-compatible bucket.
- Add `scripts/sync-maps` for syncing map files from an S3-compatible bucket.
- Add `docker-compose.production.yml` for production-specific command line options.
- Move development port bindings to `docker-compose.override.yml`.

## Notes

I created two buckets in our DigitalOcean account out-of-band. DM me on Discord and I can provide you with credentials for testing.

## Testing Instructions

See that there are no maps installed.

```bash
$ ls src/garrysmod/gamemodes/zombiesurvival/content/maps/
ls: src/garrysmod/gamemodes/zombiesurvival/content/maps/: No such file or directory
```

Run `update` which will pull our maps down and generate the FastDL content.

```bash
$ ./scripts/update
. . .
download: s3://hotdad-development-maps/zs_bar_v5.bsp to src/garrysmod/gamemodes/zombiesurvival/content/maps/zs_bar_v5.bsp
Completed 333.7 MiB/2.3 GiB (11.1 MiB/s) with 27 file(s) remaining
. . .
shasum: ./build/fastdl/maps/zs_bar_v5.bsp.bz2.sha512256: No such file or directory
Processing maps/zs_bar_v5.bsp
. . .
```

Add the following ConVars to your `src/garrysmod/cfg/server.cfg` file.

```properties
sv_downloadurl "https://hotdad-development-data.nyc3.digitaloceanspaces.com/"
sv_allowdownload "1"
sv_allowupload "1"
```

Bring up SRCDS and connect via your local instance of Garry's Mod.

```bash
$ ./scripts/server
. . .
garrysmod_1  | Logging into Steam game server account
garrysmod_1  | Connection to Steam servers successful.
garrysmod_1  |    Public IP is [redacted].
garrysmod_1  | Assigned persistent gameserver Steam ID [G-1:3528336].
garrysmod_1  | VAC secure mode is activated.
```

See that you are downloading content dependencies.

![image](https://user-images.githubusercontent.com/1774125/91794256-98664280-ebe8-11ea-9baf-e88588bad6ca.png)

See that there are no HTTP errors in the console.

![image](https://user-images.githubusercontent.com/1774125/91795400-8f2aa500-ebeb-11ea-8149-ee36329859a2.png)
